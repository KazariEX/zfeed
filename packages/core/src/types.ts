export interface Item {
    title: string;
    id?: string;
    link: string;
    date: Date;
    published?: Date;

    description?: string;
    category?: Category[];
    content?: string;

    author?: Author[];
    contributor?: Author[];
    copyright?: string;

    enclosure?: Enclosure;
    image?: string | Enclosure;
    audio?: string | Enclosure;
    video?: string | Enclosure;

    extensions?: Extension[];
}

export interface Category {
    name?: string;
    domain?: string;
    scheme?: string;
    term?: string;
}

export interface Enclosure {
    url: string;
    type?: string;
    length?: number;
    title?: string;
    duration?: number;
}

export interface Author {
    name?: string;
    email?: string;
    link?: string;
    avatar?: string;
}

export interface Extension {
  name: string;
  objects: any;
}
