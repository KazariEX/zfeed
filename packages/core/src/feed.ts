import type { Author, Extension, Item } from "./types";

export interface Feed {
    title: string;
    description?: string;
    id: string;
    link?: string;
    feed?: string;
    feedLinks?: {
        atom?: string;
        json?: string;
        rss?: string;
    };
    hub?: string;

    language?: string;
    generator?: string;
    stylesheet?: string;
    docs?: string;
    image?: string;
    favicon?: string;
    copyright?: string;
    updated?: Date;
    ttl?: number;

    category?: string;
    podcast?: boolean;

    author?: Author;
    contributors: Author[];
    categories: string[];
    items: Item[];
    extensions: Extension[];
}

export interface CreateFeedOptions extends Partial<Feed> {
    title: string;
    id: string;
}

export function createFeed(options: CreateFeedOptions) {
    const feed: Feed = {
        items: [],
        categories: [],
        contributors: [],
        extensions: [],
        ...options,
    };
    return feed;
}
