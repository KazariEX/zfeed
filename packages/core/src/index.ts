export * from "./feed";
export * from "./generators/atom1";
export * from "./generators/json1";
export * from "./generators/rss2";
export * from "./plugin";
export * from "./types";
