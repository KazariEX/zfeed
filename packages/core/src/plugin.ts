import type { Plugin } from "./types";

export function definePlugin<T>(plugin: Plugin<T>) {
    return plugin;
}
