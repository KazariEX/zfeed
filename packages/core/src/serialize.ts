export function serialize(name: string, val: unknown): string {
    if (isNullish(val)) {
        return "";
    }
    if (Array.isArray(val)) {
        return val.map((item) => serialize(name, item)).join("");
    }

    let stuff = "";
    let content = "";
    let raw = false;

    if (typeof val === "object") {
        for (const [key, value] of Object.entries(val)) {
            if (isNullish(value)) {
                continue;
            }
            if (key.startsWith("$")) {
                stuff += ` ${key.slice(1)}="${escape(String(value))}"`;
            }
            else if (key === "#text") {
                content = escape(String(value));
                raw = true;
            }
            else if (key === "#cdata") {
                content = `<![CDATA[${value}]]>`;
                raw = true;
            }
            else if (!raw) {
                content += serialize(key, value);
            }
        }
    }
    else {
        content = escape(String(val));
    }

    return name
        ? name.startsWith("?")
            ? `<?${name.slice(1)}${stuff}?>`
            : content
                ? `<${name}${stuff}>${content}</${name}>`
                : `<${name}${stuff} />`
        : content;
}

const entities: Record<string, string> = {
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    "'": "&#39;",
    "\"": "&quot;",
};

function escape(str: string) {
    return str.replaceAll(/[&<>'"]/g, (match) => entities[match]);
}

function isNullish(val: unknown) {
    return val === null || val === void 0;
}
