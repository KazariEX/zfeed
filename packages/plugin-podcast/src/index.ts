import { definePlugin } from "zfeed";
import * as packageJson from "../package.json";

/**
 * @see https://podcast-standard.org/podcast_standard
 */
export default definePlugin(() => {
    return {
        name: packageJson.name,
        resolveRss2(feed, xml) {
            xml.rss["$xmlns:googleplay"] = "http://www.google.com/schemas/play-podcasts/1.0";
            xml.rss["$xmlns:itunes"] = "http://www.itunes.com/dtds/podcast-1.0.dtd";

            const author = Array.isArray(feed.author) ? feed.author[0] : feed.author;

            if (author?.name !== void 0) {
                xml.rss.channel["googleplay:author"] = author.name;
                xml.rss.channel["itunes:author"] = author.name;
            }

            if (author?.email !== void 0) {
                xml.rss.channel["googleplay:owner"] = author.email;
                xml.rss.channel["itunes:owner"] = {
                    "itunes:email": author.email,
                };
            }

            if (feed.categories !== void 0) {
                xml.rss.channel["googleplay:category"] = feed.categories.map(({ term }) => ({ $text: term }));
                xml.rss.channel["itunes:category"] = feed.categories.map(({ term }) => ({ $text: term }));
            }

            if (feed.image !== void 0) {
                xml.rss.channel["googleplay:image"] = { $href: feed.image };
                xml.rss.channel["itunes:image"] = { $href: feed.image };
            }
        },
    };
});
