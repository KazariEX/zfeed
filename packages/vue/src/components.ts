import { h, type SetupContext } from "vue";
import type { XSLApplyImportElementAttributes, XSLApplyTemplatesElementAttributes, XSLAttributeElementAttributes, XSLAttributeSetElementAttributes, XSLCallTemplateElementAttributes, XSLChooseElementAttributes, XSLCommentElementAttributes, XSLCopyElementAttributes, XSLCopyOfElementAttributes, XSLDecimalFormatElementAttributes, XSLElementElementAttributes, XSLFallbackElementAttributes, XSLForEachElementAttributes, XSLIfElementAttributes, XSLImportElementAttributes, XSLIncludeElementAttributes, XSLKeyElementAttributes, XSLMessageElementAttributes, XSLNamespaceAliasElementAttributes, XSLNumberElementAttributes, XSLOtherwiseElementAttributes, XSLOutputElementAttributes, XSLParamElementAttributes, XSLPreserveSpaceElementAttributes, XSLProcessingInstructionElementAttributes, XSLSortElementAttributes, XSLStripSpaceElementAttributes, XSLStylesheetElementAttributes, XSLTemplateElementAttributes, XSLTextElementAttributes, XSLTransformElementAttributes, XSLValueOfElementAttributes, XSLVariableElementAttributes, XSLWhenElementAttributes, XSLWithParamElementAttributes } from "@zfeed/xsl";

function xsl<T extends Record<string, any>>(name: string) {
    return (props: T, { slots }: Omit<SetupContext, "expose">) => h(`xsl:${name}`, props, slots.default?.());
}

export const XslApplyImport = xsl<XSLApplyImportElementAttributes>("apply-import");
export const XslApplyTemplates = xsl<XSLApplyTemplatesElementAttributes>("apply-templates");
export const XslAttributeSet = xsl<XSLAttributeSetElementAttributes>("attribute-set");
export const XslAttribute = xsl<XSLAttributeElementAttributes>("attribute");
export const XslCallTemplate = xsl<XSLCallTemplateElementAttributes>("call-template");
export const XslChoose = xsl<XSLChooseElementAttributes>("choose");
export const XslComment = xsl<XSLCommentElementAttributes>("comment");
export const XslCopyOf = xsl<XSLCopyOfElementAttributes>("copy-of");
export const XslCopy = xsl<XSLCopyElementAttributes>("copy");
export const XslDecimalFormat = xsl<XSLDecimalFormatElementAttributes>("decimal-format");
export const XslElement = xsl<XSLElementElementAttributes>("element");
export const XslFallback = xsl<XSLFallbackElementAttributes>("fallback");
export const XslForEach = xsl<XSLForEachElementAttributes>("for-each");
export const XslIf = xsl<XSLIfElementAttributes>("if");
export const XslImport = xsl<XSLImportElementAttributes>("import");
export const XslInclude = xsl<XSLIncludeElementAttributes>("include");
export const XslKey = xsl<XSLKeyElementAttributes>("key");
export const XslMessage = xsl<XSLMessageElementAttributes>("message");
export const XslNamespaceAlias = xsl<XSLNamespaceAliasElementAttributes>("namespace-alias");
export const XslNumber = xsl<XSLNumberElementAttributes>("number");
export const XslOtherwise = xsl<XSLOtherwiseElementAttributes>("otherwise");
export const XslOutput = xsl<XSLOutputElementAttributes>("output");
export const XslParam = xsl<XSLParamElementAttributes>("param");
export const XslPreserveSpace = xsl<XSLPreserveSpaceElementAttributes>("preserve-space");
export const XslProcessingInstruction = xsl<XSLProcessingInstructionElementAttributes>("processing-instruction");
export const XslSort = xsl<XSLSortElementAttributes>("sort");
export const XslStripSpace = xsl<XSLStripSpaceElementAttributes>("strip-space");
export const XslStylesheet = xsl<XSLStylesheetElementAttributes>("stylesheet");
export const XslTemplate = xsl<XSLTemplateElementAttributes>("template");
export const XslText = xsl<XSLTextElementAttributes>("text");
export const XslTransform = xsl<XSLTransformElementAttributes>("transform");
export const XslValueOf = xsl<XSLValueOfElementAttributes>("value-of");
export const XslVariable = xsl<XSLVariableElementAttributes>("variable");
export const XslWhen = xsl<XSLWhenElementAttributes>("when");
export const XslWithParam = xsl<XSLWithParamElementAttributes>("with-param");
