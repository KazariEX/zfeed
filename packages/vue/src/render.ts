import { renderToString } from "@vue/server-renderer";
import { type Component, createSSRApp } from "vue";

/**
 * Copy from `vue-component-type-helpers`
 *
 * @link https://github.com/vuejs/language-tools/blob/v2.2.10/packages/component-type-helpers/index.ts#L6
 */
type ComponentProps<T> = T extends new (...args: any) => {
    $props: infer P;
} ? NonNullable<P> : T extends (props: infer P, ...args: any) => any ? P : {};

export async function render<T extends Component>(component: T, props?: ComponentProps<T>) {
    const app = createSSRApp(component, props);
    const text = await renderToString(app);
    return `<?xml version="1.0" encoding="UTF-8"?>${text}`;
}
