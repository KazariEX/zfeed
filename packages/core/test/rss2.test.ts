import { describe, expect, it } from "vitest";
import { generateRss2 } from "../src";
import { feed, format } from "./shared";

describe("rss2", () => {
    it("snapshot", () => {
        const result = format(generateRss2(feed));
        expect(result).toMatchSnapshot();
    });
});
