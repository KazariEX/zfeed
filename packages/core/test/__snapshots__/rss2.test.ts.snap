// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rss2 > snapshot 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<?xml-stylesheet type="text/xsl" href="http://example.com/stylesheet.xsl"?>
<rss version="2.0" xml:lang="en-US" xmlns:media="http://search.yahoo.com/mrss/" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:content="http://purl.org/rss/1.0/modules/content/">
  <channel>
    <title>Feed Title</title>
    <description>This is my personnal feed!</description>
    <link>http://example.com/</link>
    <lastBuildDate>Mon, 30 Sep 2019 11:45:14 GMT</lastBuildDate>
    <category>Anime</category>
    <category>Novel</category>
    <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
    <generator>https://example.com/generator</generator>
    <language>en-US</language>
    <ttl>60</ttl>
    <image>
      <title>Feed Title</title>
      <url>http://example.com/image.png</url>
      <link>http://example.com/</link>
    </image>
    <copyright>Copyright Â© 2025 KazariEX</copyright>
    <atom:link href="http://example.com/feed" rel="self" type="application/rss+xml"></atom:link>
    <atom:link href="wss://example.com/hub" rel="hub"></atom:link>
    <item>
      <title>Hello World</title>
      <guid isPermaLink="false">https://example.com/hello-world?id=this&amp;that=true</guid>
      <link>https://example.com/hello-world</link>
      <pubDate>Fri, 19 Jul 2019 11:45:14 GMT</pubDate>
      <description>
        <![CDATA[This is an article about Hello World.]]>
      </description>
      <category>Category One</category>
      <category domain="http://example.com/category">Category Two</category>
      <content:encoded>
        <![CDATA[Content of the item.]]>
      </content:encoded>
      <author>mimori@example.com (Mimori)</author>
      <author>yamabuki@example.com (Yamabuki)</author>
      <enclosure url="https://example.com/hello-world.jpg" length="0" type="image/jpg"></enclosure>
      <extension:1>
        <about>just an item extension example</about>
        <dummy>example</dummy>
      </extension:1>
      <extension:2>
        <about>just a second item extension example</about>
        <dummy>example</dummy>
      </extension:2>
    </item>
  </channel>
  <extension>
    <about>just an extension example</about>
    <dummy>example</dummy>
  </extension>
</rss>
"
`;
